### PR TITLE
add return_token_timestamps to WhisperProcessor

### LINKED ISSUE #30433: Timestamps are broken for whisper large with WhisperForConditionalGeneration
### System Info

**System Info**

- `transformers` version: 4.40.0.dev0
- Platform: Linux-5.4.0-166-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.22.2
- Safetensors version: 0.4.2
- Accelerate version: 0.29.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.2.2+cu121 (True)
- Tensorflow version (GPU?): 2.13.1 (True)
- Flax version (CPU?/GPU?/TPU?): 0.7.0 (cpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

**Who can help**

@kamilakesbi @sanchit-gandhi 

**Reproduction**

Timestamps are broken for whisper-large-v3 when used with  `WhisperForConditionalGeneration`: 

Note: This issue is related to #30224 

### Reproduction

```
from transformers import WhisperProcessor, WhisperForConditionalGeneration
import torch 
from datasets import load_dataset

device = "cuda:0" if torch.cuda.is_available() else "cpu"

processor = WhisperProcessor.from_pretrained("openai/whisper-large-v3")
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-large-v3")
model.to(device)

ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
speech_samples = ds.sort("id").select(range(4))[:4]["audio"]

input_speech = [x["array"] for x in speech_samples]
features = processor.feature_extractor(raw_speech=input_speech, return_tensors="pt")

input_features = features.input_features.to(device)
generate_kwargs = {}

generate_outputs = model.generate(
    input_features, return_timestamps=True, return_token_timestamps=True, **generate_kwargs
)
print(generate_outputs.token_timestamps)
```
We get: 

```
tensor([[ 0.0000,  0.0000, 29.3000, 29.3000, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800],
        [ 0.0000,  0.0000, 29.3000, 29.3000, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800, 29.9800,
         29.9800],...
```
indicating that the timestamps are broken...

### PR DESCRIPTION
**Pull request by kamilakesbi**
# What this PR do ? 

This PR fixes #30433 by making sure we can compute timestamps with both `WhisperForConditionalGeneration` and `AutomaticSpeechRecognitionPipeline`. 

We add a `return_timestamps` hyperparameter to `WhisperProcessor.feature_extractor` to be used when we want to compute timestamps.  When True, the processor will return a `num_frames` parameter containing the number of frames of the input audios. `num_frames` is then passed to `generate` and used to compute timestamps. 

Prior to that, timestamps were broken for whisper-large-v3 when used with WhisperForConditionalGeneration. 

# Who can review ? 

cc @sanchit-gandhi

### CODE REVIEW

**Review by sanchit-gandhi — COMMENTED** [review]
Think this is indeed the cleanest and most reliable approach for computing `num_frames`. The alternative method we discussed offline is detailed below. Leaving it here for the next reviewer to consider, in case they believe it's a superior strategy.

Anything beyond `len(input_speech)` is padded by zeros to 30-seconds in the feature extractor. If we know what zero’s correspond to in log-mel space, then we can know how many padded zeros we have in our spectrogram, and thus what the original input length was. 

Note that this won’t be perfect: the last frame where the audio stops is going to be affected by the end of the audio, so we’ll be looking for the first frame where there is entirely padding (rather than finding the frame in which the audio stops). 

However, the original method by OpenAI (and the one implemented in this PR) is also imperfect: if a user took a 10-second audio, and padded it by hand to 15-seconds with zeros, then `num_frames` would be computed on the length of the padded input, not the original one

File: src/transformers/models/whisper/generation_whisper.py [hunk 1]
```diff
@@ -474,6 +475,13 @@ def generate(
                 "The input name `inputs` is deprecated. Please make sure to use `input_features` instead.",
                 FutureWarning,
             )
+
+        if input_features is not None and isinstance(input_features, BatchFeature):
```
[the remark this answers is not in the document]
Comment (kamilakesbi) [reply src/transformers/models/whisper/generation_whisper.py:479]: ok I see, shouldn't we use `processor.feature_extractor` ?  in this case the output is a `BatchFeature`.

[the remark this answers is not in the document]
Comment (kamilakesbi) [reply src/transformers/models/whisper/generation_whisper.py:479]: The output of the processor here would be a `BatchFeature` as indicated [here](https://github.com/huggingface/transformers/blob/3f435823e032666aa9198c304774d138c541eb1c/src/transformers/models/whisper/feature_extraction_whisper.py#L192) no ?

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 2]
```diff
@@ -237,6 +238,9 @@ def __call__(
             device (`str`, *optional*, defaults to `'cpu'`):
                 Specifies the device for computation of the log-mel spectrogram of audio signals in the
                 `_torch_extract_fbank_features` method. (e.g., "cpu", "cuda")
+            return_token_timestamps (`bool`, *optional*, defaults to None):
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/feature_extraction_whisper.py:241]:
```suggestion
            return_token_timestamps (`bool`, *optional*, defaults to `None`):
```

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 3]
```diff
@@ -312,4 +317,9 @@ def __call__(
         if return_tensors is not None:
             padded_inputs = padded_inputs.convert_to_tensors(return_tensors)
 
+        if return_token_timestamps is not None:
+            padded_inputs["num_frames"] = torch.tensor(
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/feature_extraction_whisper.py:321]: This should use `.convert_to_tensor` to respect the tensor type requested by the user, not forcing it to a torch tensor

File: src/transformers/models/whisper/generation_whisper.py [hunk 4]
```diff
@@ -207,7 +208,7 @@ def _extract_token_timestamps(self, generate_outputs, alignment_heads, time_prec
             # two cases:
             # 1. num_frames is the same for each sample -> compute the DTW matrix for each sample in parallel
             # 2. num_frames is different, compute the DTW matrix for each sample sequentially
-
+            num_frames = num_frames.tolist() if isinstance(num_frames, torch.Tensor) else num_frames
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:211]: Is it not more efficient to keep `num_frames` as a tensor if possible?

File: src/transformers/models/whisper/generation_whisper.py [hunk 5]
```diff
@@ -231,7 +232,7 @@ def _extract_token_timestamps(self, generate_outputs, alignment_heads, time_prec
 
         # Perform dynamic time warping on each element of the batch.
         for batch_idx in range(batch_size):
-            if num_frames is not None and isinstance(num_frames, (tuple, list, np.ndarray)):
+            if num_frames is not None and isinstance(num_frames, (tuple, list, np.ndarray, torch.Tensor)):
                 matrix = weights[batch_idx, ..., : num_frames[batch_idx] // 2]
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:236]: Since here we slice `weights` (torch tensor) by `num_frames` (which should be most efficient if both are tensors no?)

File: src/transformers/models/whisper/generation_whisper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/generation_whisper.py:479]: Yes, but then we un-pack the `BatchFeature` when we pass it to the model, i.e. we do:
```python
pred_ids = model.generate(**inputs)
```
Not:
```python
pred_ids = model.generate(inputs)
```

File: tests/models/whisper/test_modeling_whisper.py [hunk 6]
```diff
@@ -1912,6 +1912,177 @@ def test_tiny_timestamp_generation(self):
...
+
+        input_speech = np.concatenate(self._load_datasamples(4))
+        input_features = processor(
+            input_speech, return_tensors="pt", sampling_rate=16_000, return_token_timestamps=True
+        ).input_features
+        input_features = input_features.to(torch_device)
+
+        generated_ids = model.generate(input_features, max_length=448, return_timestamps=True).to("cpu")
+
+        EXPECTED_OUTPUT = torch.tensor(
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_whisper.py:1930]: Can you put the expected outputs on one line, and add ` # fmt: skip ` at the end?
```
EXPECTED_OUTPUT = torch.tensor([ 50258, ..., 50257]) # fmt: skip
```

E.g. as done here:
https://github.com/huggingface/transformers/blob/f4014e75db0190792b3feeccfc5dc5b5f9f0ce7b/tests/models/whisper/test_modeling_whisper.py#L1903

This will stop the linter from formatting it over hundreds of lines, thus improving readability of the test

File: tests/models/whisper/test_modeling_whisper.py [hunk 7]
```diff
@@ -1912,6 +1912,177 @@ def test_tiny_timestamp_generation(self):
...
+                10281,
+                934,
+                439,
+                11,
+                51799,
+                51815,
+                50257,
+            ]
+        )
+        self.assertTrue(torch.allclose(generated_ids, EXPECTED_OUTPUT))
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_whisper.py:2041]: `.allclose` doesn't really make sense for ints?

```suggestion
        self.assertTrue((generated_ids == EXPECTED_OUTPUT).all())
```

**Review by sanchit-gandhi — APPROVED** [review]
Looks good! Mostly formatting now, then we can get a final review

File: src/transformers/models/whisper/generation_whisper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kamilakesbi) [reply src/transformers/models/whisper/generation_whisper.py:479]: If we un-pack we also loose `num_frames` here.

[the remark this answers is not in the document]
Comment (kamilakesbi) [reply src/transformers/models/whisper/generation_whisper.py:479]: In this case it will work with both packed and unpacked inputs. Isn't that better?

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 8]
```diff
@@ -302,13 +306,17 @@ def __call__(
...
+
         else:
             padded_inputs["input_features"] = input_features
 
         if return_attention_mask:
             # rescale from sample (48000) to feature (3000)
             padded_inputs["attention_mask"] = padded_inputs["attention_mask"][:, :: self.hop_length]
 
+        if return_token_timestamps is not None:
+            padded_inputs["num_frames"] = [len(raw_speech[i]) // self.hop_length for i in range(len(raw_speech))]
```
Comment (amyeroberts) [inline src/transformers/models/whisper/feature_extraction_whisper.py:318]: `range(len(x))` is a bit of a code smell in python - there's no need to index when you can iterate over objects directly 

```suggestion
            padded_inputs["num_frames"] = [len(raw_speech_i) // self.hop_length for raw_speech_i in raw_speech]
```

File: src/transformers/models/whisper/generation_whisper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/whisper/generation_whisper.py:482]: I'm aligned with @sanchit-gandhi here - handling packed and unpacked inputs isn't something any of our other processing classes handle, so it's not something we need to introduce here

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this feature and tests! 

All looks good to me - just the handling of unpacked features to remove

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

File: tests/models/whisper/test_modeling_whisper.py [hunk 9]
```diff
@@ -1927,7 +1927,117 @@ def test_large_timestamp_generation(self):
 
         generated_ids = model.generate(input_features, max_length=448, return_timestamps=True).to("cpu")
 
-        EXPECTED_OUTPUT = torch.tensor([50258, 50259, 50360, 50365, 2221, 13, 2326, 388, 391, 307, 264, 50244, 295, 264, 2808, 5359, 11, 293, 321, 366, 5404, 281, 2928, 702, 14943, 13, 50629, 50682, 6966, 307, 2221, 13, 2326, 388, 391, 311, 9060, 1570, 1880, 813, 702,  1871, 13, 50870, 50911, 634, 5112, 505, 300, 412, 341, 42729, 3196, 295, 264,  1064,  11, 365,  5272,   293, 12904,  9256, 450, 10539, 949, 505, 11, 51245, 51287,  1034, 4680, 10117, 490, 3936, 293, 1080,  3542, 5160, 881, 26336, 281, 264, 1575, 13, 51494, 51523, 634, 575, 12525, 22618, 1968,  6144, 35617, 1456, 397, 266, 311, 589, 307, 534, 10281, 934, 439, 11, 51799, 51815, 50257])
+        EXPECTED_OUTPUT = torch.tensor(
```
Comment (amyeroberts) [inline tests/models/whisper/test_modeling_whisper.py:1930]: I don't think we want to split across the lines like this. 

You can wrap `EXPECTED_OUTPUT` around `# fmt: off` and `fmt: on` comments to avoid this

Comment (kamilakesbi) [reply tests/models/whisper/test_modeling_whisper.py]: Ok thanks for the tips! will be useful ;)

Comment (sanchit-gandhi) [reply tests/models/whisper/test_modeling_whisper.py]: You can also use `# fmt: skip` for single lines, _c.f._ the previous comment https://github.com/huggingface/transformers/pull/30812#discussion_r1603595554

**Comment by kamilakesbi** [comment]
cc @amyeroberts @sanchit-gandhi Could you please merge this PR as I don't have the rights to do so? 
Thanks!
